apps/ex_nvr/lib/ex_nvr/model/remote_storage.ex

```diff
@@ -0,0 +1,123 @@
+defmodule ExNVR.Model.RemoteStorage do
+  use Ecto.Schema
+
+  alias Ecto.Changeset
+
+  @type t :: %__MODULE__{
+          type: :s3 | :http,
+          config: Config.t() | nil,
+          inserted_at: DateTime.t() | nil,
+          updated_at: DateTime.t() | nil
+        }
+
+  defmodule Config do
+    use Ecto.Schema
+
+    alias Ecto.Changeset
+
+    @type t :: %__MODULE__{
+            url: binary(),
+            username: binary(),
+            password: binary(),
+            token: binary(),
+            bucket: binary(),
+            access_key_id: binary(),
+            secret_access_key: binary(),
+            region: binary()
+          }
+
+    @primary_key false
+    embedded_schema do
+      field :url, :string
+      field :username, :string
+      field :password, :string
+      field :token, :string
+      field :bucket, :string
+      field :access_key_id, :string
+      field :secret_access_key, :string
+      field :region, :string
+    end
+
+    def changeset(struct, params, :s3) do
+      struct
+      |> Changeset.cast(params, [
+        :url,
+        :bucket,
+        :access_key_id,
+        :secret_access_key,
+        :region
+      ])
+      |> Changeset.validate_required([
+        :bucket,
+        :access_key_id,
+        :secret_access_key,
+        :region
+      ])
+      |> Changeset.validate_change(:url, fn :url, url -> validate_url(:url, url) end)
+    end
+
+    def changeset(struct, params, :http) do
+      struct
+      |> Changeset.cast(params, [
+        :url,
+        :username,
+        :password,
+        :token
+      ])
+      |> Changeset.validate_required([:url])
+      |> Changeset.validate_change(:url, fn :url, url -> validate_url(:url, url) end)
+    end
+
+    defp validate_url(field, url) do
+      parsed_url = URI.parse(url)
+
+      cond do
+        parsed_url.scheme not in ["http", "https"] ->
+          [{field, "scheme should be http or https"}]
+
+        to_string(parsed_url.host) == "" ->
+          [{field, "invalid #{parsed_url.scheme} url"}]
+
+        true ->
+          []
+      end
+    end
+  end
+
+  schema "remote_storages" do
+    field :name, :string
+    field :type, Ecto.Enum, values: [:s3, :http], default: :s3
+
+    embeds_one :config, Config, on_replace: :update
```

Comment:
Let's experiment with using multiple embedded schema with the same source field in database
```elixir
schema "remote_storages"
  field :name, :string
  field :type, ...
  field :url, :string
  
  embeds_one :s3_config, S3Config, souce: :config, on_replace: :update
  embeds_one :http_config, HttpConfig, source: :config, on_replace: :update
...
```
Please try it. It'll make the config/changeset more readable.